pkg/querier/queryrange/downstreamer.go

```diff
@@ -249,3 +258,374 @@ func ResponseToResult(resp queryrangebase.Response) (logqlmodel.Result, error) {
 		return logqlmodel.Result{}, fmt.Errorf("cannot decode (%T)", resp)
 	}
 }
+
+// downstreamAccumulator is one of two variants:
+// a logsAccumulator or a bufferedAccumulator.
+// Which variant is detected on the first call to Accumulate.
+// Metric queries, which are generally small payloads, are buffered
+// since the memory overhead is negligible.
+// Log queries, sharded thousands of times and each returning <limit>
+// results, can be _considerably_ larger.	In this case, we eagerly
+// accumulate the results into a logsAccumulator, discarding values
+// over the limit to keep memory pressure down while other subqueries
+// are executing.
+type downstreamAccumulator struct {
+	acc    resultAccumulator
+	params logql.Params
+	n      int // number of queries, used to build slice size
+}
+
+type resultAccumulator interface {
+	Accumulate(logqlmodel.Result, int) error
+	Result() []logqlmodel.Result
+}
+
+func newDownstreamAccumulator(params logql.Params, nQueries int) *downstreamAccumulator {
+	return &downstreamAccumulator{params: params, n: nQueries}
+}
+
+func (a *downstreamAccumulator) build(acc logqlmodel.Result) {
+	if acc.Data.Type() == logqlmodel.ValueTypeStreams {
+
+		// the stream accumulator stores a heap with reversed order
+		// from the results we expect, so we need to reverse the direction
+		direction := logproto.FORWARD
+		if a.params.Direction() == logproto.FORWARD {
+			direction = logproto.BACKWARD
+		}
+
+		a.acc = newStreamAccumulator(direction, int(a.params.Limit()))
+
+	} else {
+		a.acc = &bufferedAccumulator{
+			results: make([]logqlmodel.Result, a.n),
+		}
+
+	}
+}
+
+func (a *downstreamAccumulator) Accumulate(ctx context.Context, index int, acc logqlmodel.Result) error {
```

Comment:
Is this more of an append or add?

Reply:
it's unknown if we'll actually accept any samples at this point. I think accumulate is the right word.